runtime/lua/lsp/structures.lua

```diff
@@ -0,0 +1,183 @@
+  return args
+    or __version
+end
+structures.text = function(args)
+  return args
+    or lsp_util.get_buffer_text(0)
+end
+structures.TextDocumentIdentifier = function(args)
+  args = check_table(args)
+
+  return {
+    uri = structures.DocumentUri(args.uri),
+  }
+end
+structures.VersionedTextDocumentIdentifier = function(args)
+  args = check_table(args)
+
+  local result = structures.TextDocumentIdentifier(args)
+  result.version = structures.version(args.version)
+
+  return result
+end
+structures.TextDocumentItem = function(args)
+  args = check_table(args)
+
+  return {
+    uri = structures.DocumentUri(args.uri),
+    languageId = structures.languageId(args.languageId),
+    version = structures.version(args.version),
+    text = structures.text(args.text),
+  }
+end
+structures.line = function(args)
+  return args
+    -- TODO: Check the conversion of some of these functions from nvim <-> lua
+    or (vim.api.nvim_call_function('line', {'.'}) - 1)
+end
+structures.character = function(args)
+  return args
+    or (vim.api.nvim_call_function('col', {'.'}) - 1)
+end
+
+structures.Position = function(args)
+  args = check_table(args)
+
+  return {
+    line = structures.line(args.line),
+    character = structures.character(args.character),
+  }
+end
+structures.TextDocumentPositionParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.TextDocumentIdentifier(args.textDocument),
+    position = structures.Position(args.position),
+  }
+end
+structures.ReferenceContext = function(args)
+  args = check_table(args)
+
+  return {
+    includeDeclaration = args.includeDeclaration or true,
+  }
+end
+structures.ReferenceParams = function(args)
+  args = check_table(args)
+
+  local positionParams = structures.TextDocumentPositionParams(args)
+  positionParams.context = structures.ReferenceContext(args.context)
+
+  return positionParams
+end
+structures.DidOpenTextDocumentParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.TextDocumentItem(args.textDocument)
+  }
+end
+
+structures.DidSaveTextDocumentParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.TextDocumentItem(args.textDocument),
+    text = table.concat(vim.api.nvim_buf_get_lines(0, 0, -1, false), "\n"),
+  }
+end
+
+structures.WillSaveTextDocumentParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.TextDocumentItem(args.textDocument),
+    reason = args.reason or protocol.TextDocumentSaveReason.Manual,
+  }
+end
+
+-- TODO: Incremental changes.
+--  Maybe use the PR that externalizes that once its merged
+structures.DidChangeTextDocumentParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.VersionedTextDocumentIdentifier(args.textDocument),
+    contentChanges = {
+      { text = table.concat(vim.api.nvim_buf_get_lines(0, 0, -1, false), "\n") },
```

Comment:
if `nvim_buf_get_option(0, 'eol')` is true (the typical case), an extra `"\n"` should be added at the end.